inc/render.h

```diff
@@ -0,0 +1,171 @@
+/*                                                                            */
+/* ************************************************************************** */
+
+#ifndef RENDER_H
+# define RENDER_H
+
+/* ====== LIBRARIES ====== */
+
+# include <mlx.h>
+# include <X11/X.h>
+
+# include "error_mlx.h"
+# include "camera.h"
+# include "ft_print.h"
+# include "cleanup.h"
+# include "entities.h"
+# include "key_codes.h"
+
+/* ====== TYPEDEFS ====== */
+
+/**
+ * @brief mlx img struct.
+ *
+ * @param ptr		Pointer to mlx img.
+ * @param addr		Pointer to the char *buffer of the mlx img.
+ * @param bpp		Bytes per pixel.
+ * @param line_len	Line length.
+ * @param endian	Endian.
+ * @param bytes		Number of bytes per pixel.
+ *
+ */
+typedef struct s_img
+{
+	void	*ptr;
+	char	*addr;
+	int		bpp;
+	int		line_len;
+	int		endian;
+	int		bytes;
+}	t_img;
+
+/**
+ * @brief mlx pointers struct.
+ *
+ * @param mlx_ptr	Pointer to mlx.
+ * @param win_ptr	Pointer to mlx window.
+ * @param img		mlx img struct.
+ */
+typedef struct s_mlx_ptrs
+{
+	void	*mlx_ptr;
+	void	*win_ptr;
+	t_img	img;
+}	t_mlx_ptrs;
+
+/**
+ * @brief Overarching render struct.
+ *
+ * @param mlx_ptrs		mlx pointers struct.
+ * @param cam			Camera struct.
+ * @param scene			Scene struct.
+ * @param image			Image struct.
+ * @param vp			Viewport struct.
+ */
+typedef struct s_render
+{
+	t_mlx_ptrs	mlx_ptrs;
+	t_cam		cam;
+	t_entities	scene;
+	t_image		image;
+	t_viewport	vp;
+}	t_render;
+
+/* ====== FUNCTIONS ====== */
+
+// render_init.c
+
+/**
+ * @brief Initializes the mlx_ptrs struct.
+ *
+ * @param mlx_ptrs		Pointer to mlx_ptrs struct.
+ * @param image			Pointer to image struct.
+ * @param fullscreen	Fullscreen flag.
+ * @return t_err		SUCCESS, ERROR.
+ */
+t_err	ft_init_mlx_ptrs(t_mlx_ptrs *mlx_ptrs, t_image *image, bool fullscreen);
+
+/**
+ * @brief Initializes the mlx img struct.
+ *
+ * @param mlx_ptrs		Pointer to mlx_ptrs struct.
+ * @param size			Image size as int array.
+ * @return t_err		SUCCESS, ERROR.
+ */
+t_err	ft_init_image(t_mlx_ptrs *mlx_ptrs, int size[2]);
+
+/**
+ * @brief Sets the window to fullscreen.
+ *
+ * @param mlx_ptrs	Pointer to mlx_ptrs struct.
+ * @param image		Pointer to image struct.
+ */
+void	ft_set_fullscreen(t_mlx_ptrs *mlx_ptrs, t_image *image);
+
+// render_draw.c
+
+/**
+ * @brief Draws a single pixel to the mlx img.
+ *
+ * Loops for the number of bytes per pixel and writes the colour to the address.
+ * @param img	Pointer to mlx img struct.
+ * @param x		X coordinate.
+ * @param y		Y coordinate.
+ * @param color	Colour to write.
+ */
+void	ft_put_pix_to_image(t_img *img, int x, int y, int color);
+
+/**
+ * @brief Renders the scene to the mlx img.
+ *
+ * Creates rays for each pixel of the image and "shoots" them
+ * through the corresponding pixel in the viewport. The colour of
+ * each ray is returned and written to the mlx img.
+ * @param render	Pointer to render struct.
```

Comment:
This was an oversight, but also addresses a spicy topic: is there a difference between passing by value or passing by pointer?

The highest rated answer on SO thread from 2020 says with no optimizations it should be the same. But with optimizations it makes a difference: [Link to Thread](https://stackoverflow.com/questions/59976295/structs-and-unions-which-is-better-from-a-performance-point-of-view-passing-th)
Interestingly enough another one argues that the performance gain is negible on newer machines, so that it only makes a difference on embedded or mobile.

On the ACCU21 conference there was a talk about Modern C: [Link to YouTube Video](https://youtu.be/QpAhX-gsHMs?si=Eb_RAOE4TbBZdKhV)
The speaker states, that passing by value is generally more efficient nowadays, than passing by pointer (which contradicts the thread above).

Right now we init t_vec3 structs via value return.
We generally pass pointers of structs to functions but also sometimes not (e.g. ft_initiate_viewport() )
I wouldn't touch the t_vec3 fts since they are nice to work with. But I think we should uniformly pass by value or pointer 

A pointer is 8byte in size and our structs are all bigger than that. Therefore pointer passing is generally lighter. Going by SO passing by pointer should be better with optimizations.
Since in base we're not "allowed" to compile with optimizations - at least for base subject - do we care about optimizations on this level? 
The advantage of always passing by value would be that it is easier to write (render.cam instead of render->cam).

Reply:
I would go with pointers as this should be in theory more efficient, especially if the content of that pointer supersedes it by far in size. Also it keeps the passed size nicely uniform.

It may not make a difference in the base version, but then again, does it really make a difference at all? One that we would notice? On our machines the calculation power is that immense that I doubt that the jump from address to value results in a noticeable difference..

Reply:
Maybe it was not clear: I don't know which one is better, since the two sources say the exact opposite. 
Passing by pointer should be faster would also be my first intuition since we don't need to allocate the whole struct on the stack. 
However passing by value seems to be highly optimized anyway --> maybe its only true for newer standards like C11 or C20?

I think we shouldn't go for which is faster but which is more readable.
And here passing by value wins in my book, since we don't need the arrow operator. This makes it shorter and for me more readable. Even more so when a subelement is addressed:

`render.cam.look_at` vs `render->cam.look_at`

Reply:
I don't see no issue in readability with the pointers. 
Regarding efficiency: if there is a debate going on between the fractions, we are certainly not (yet) in a position to find a solution. So it will be more a gut decision, and my gut also tells me that pointers should be better.

But then again, I don't think it really matters

Reply:
I updated ft_draw_scene and initiate_viewport.

I checked and you'd need to update most of the raytracer functions. Almost all of them pass by value (ray.h, lighting.h, hit_manager.h). Currently pass by pointer is only used if you want to change the thing in the function.